Expecto/Expecto.Impl.fs

```diff
@@ -559,8 +580,63 @@ module Impl =
               }
       }
 
+  let inline internal setStatus (status : ActivityStatusCode) (span : Activity) =
+    if isNull span |> not then
+      span.SetStatus(status) |> ignore
+
+  let inline internal setExn (e : exn) (span : Activity) =
+    if isNull span |> not then
+      let tags =
+          ActivityTagsCollection(
+              seq {
+                  KeyValuePair("exception.type", box (e.GetType().Name))
+                  KeyValuePair("exception.stacktrace", box (e.ToString()))
+                  if not <| String.IsNullOrEmpty(e.Message) then
+                      KeyValuePair("exception.message", box e.Message)
+              }
+          )
+
+      ActivityEvent("exception", tags = tags)
+      |> span.AddEvent
+      |> ignore
+
+  let inline internal setExnMarkFailed (e : exn) (span : Activity) =
+    if isNull span |> not then
+      setExn e span
+      span  |> setStatus ActivityStatusCode.Error
+
+  let setSourceLocation (sourceLoc : SourceLocation) (span : Activity) =
+    if isNull span |> not && sourceLoc <> SourceLocation.empty then
+      span.SetTag("code.lineno", sourceLoc.lineNumber) |> ignore
+      span.SetTag("code.filepath", sourceLoc.sourcePath) |> ignore
+
+  let inline internal addOutcome (result : TestResult) (span : Activity) =
+    if isNull span |> not then
+      span.SetTag("test.result.status", result.tag) |> ignore
+      span.SetTag("test.result.message", result) |> ignore
+
+  let inline internal  start (span : Activity) =
+    if isNull span |> not then
+      span.Start() |> ignore
+    span
+
+  let inline internal stop (span : Activity) =
+    if isNull span |> not then
+      span.Stop() |> ignore
+
+  let inline internal createActivity (name : string) (source : ActivitySource option) =
+    match source with
+    | Some source when not(isNull source) -> source.CreateActivity(name, ActivityKind.Internal)
+    | _ -> null
+
   let execTestAsync (ct:CancellationToken) config (test:FlatTest) : Async<TestSummary> =
     async {
+      let span =
+        config.activitySource
+        |> createActivity (config.joinWith.format test.name)
```

Comment:
This is the starting point for creating an span. Later we add additional information to it based on whether we pass/fail/ignore.